src/poetry/puzzle/provider.py

```diff
@@ -221,6 +221,42 @@ def search_for_installed_packages(
         )
         return packages
 
+    def search_for_direct_origin_dependency(self, dependency: Dependency) -> Package:
+        cached = self._deferred_cache.get(dependency)
+        if cached is not None:
```

Comment:
Should we also update dependency._constraint and dependency._pretty_constraint here?

Reply:
oh dear, that was even half of my point in the first place...